pkg/backends/alb/mech/nlm/newest_last_modified.go

```diff
@@ -81,111 +84,84 @@ func (h *handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-	contexts := make([]context.Context, sz)
-	for i := range sz {
-		contexts[i] = context.Background()
+	// Create contexts for cancellation
+	contexts := make([]context.Context, l)
+	cancels := make([]context.CancelFunc, l)
+	for i := range l {
+		contexts[i], cancels[i] = context.WithCancel(r.Context())
 	}
-	nrm := &newestResponseMux{i: -1, contexts: contexts}
-	nrm.wg.Add(sz)
-	return nrm
-}
 
-func (nrm *newestResponseMux) registerLM(i int, t time.Time) bool {
-	var ok bool
-	if t.IsZero() {
-		return false
-	}
-	if nrm.t.Load().IsZero() || t.After(nrm.t.Load()) {
-		atomic.StoreInt64(&nrm.i, int64(i))
-		nrm.t.Store(t)
-		ok = true
-	}
-	return ok
-}
+	// Track the newest Last-Modified response
+	newestIdx := -1
+	var newestTime time.Time
+	var mu sync.Mutex
 
-func (nrm *newestResponseMux) getNewest() int64 {
-	return atomic.LoadInt64(&nrm.i)
-}
+	// Capture all responses
+	captures := make([]*capture.CaptureResponseWriter, l)
+	var wg sync.WaitGroup
 
-func newNewestResponseGate(w http.ResponseWriter, i int,
-	nrm *newestResponseMux,
-) *newestResponseGate {
-	return &newestResponseGate{
-		ResponseWriter: w, h: http.Header{},
-		i: int64(i), nrm: nrm,
+	// Fanout to all healthy targets
+	for i := range l {
+		if hl[i] == nil {
+			continue
+		}
+		idx := i
+		wg.Go(func() {
+			r2, _ := request.Clone(r)
+			r2 = request.ClearResources(r2.WithContext(contexts[idx]))
+			crw := capture.NewCaptureResponseWriter()
+			captures[idx] = crw
+			hl[idx].ServeHTTP(crw, r2)
+
+			if lmStr := crw.Header().Get(headers.NameLastModified); lmStr != "" {
+				lm, err := time.Parse(time.RFC1123, lmStr)
+				if err == nil && !lm.IsZero() {
+					mu.Lock()
+					if newestIdx == -1 || lm.After(newestTime) {
+						newestIdx = idx
+						newestTime = lm
+					}
+					mu.Unlock()
+				}
+			}
+		})
 	}
-}
 
-func (nrg *newestResponseGate) Header() http.Header {
-	return nrg.h
-}
-
-func (nrg *newestResponseGate) WriteHeader(i int) {
-	nrg.s = int64(i)
-	nrg.wh = nrg.h
-	nrg.h = nil
-	lm, err := time.Parse(time.RFC1123, nrg.wh.Get(headers.NameLastModified))
-	if err == nil {
-		nrg.ca = !nrg.nrm.registerLM(int(nrg.i), lm)
-	}
-	nrg.nrm.wg.Done()
-}
+	// Wait for all responses to complete
+	wg.Wait()
 
-func (nrg *newestResponseGate) Write(b []byte) (int, error) {
-	if nrg.ca { // can abort without waiting, since this gate is already proven
-		// not to be newest
-		return len(b), nil
-	}
-	nrg.nrm.wg.Wait()
-	if nrg.nrm.getNewest() == nrg.i {
-		if len(nrg.wh) > 0 {
-			headers.Merge(nrg.ResponseWriter.Header(), nrg.wh)
-			nrg.wh = nil
+	// Write the response with the newest Last-Modified
+	if newestIdx >= 0 && newestIdx < len(captures) && captures[newestIdx] != nil {
+		crw := captures[newestIdx]
+		headers.Merge(w.Header(), crw.Header())
+		statusCode := crw.StatusCode()
+		w.WriteHeader(statusCode)
+		w.Write(crw.Body())
+
+		// Cancel all other contexts
+		for j, cancel := range cancels {
+			if j != newestIdx {
+				cancel()
+			}
+		}
```

Comment:
i don't think this cancel is necessary (here and in the next branch around L160)

on L132 we are waiting for all responses to complete. Is there anything to cancel at this point, aren't the contexts done being used?